src/berryimu.py

```diff
@@ -0,0 +1,243 @@
+
+    async def get_properties(self, *, extra: Optional[Dict[str, Any]] = None, timeout: Optional[float] = None, **kwargs) -> MovementSensor.Properties:
+        return MovementSensor.Properties(linear_velocity_supported=False,
+                                         angular_velocity_supported=True,
+                                         orientation_supported=True,
+                                         position_supported=False,
+                                         compass_heading_supported=True,
+                                         linear_acceleration_supported=True)
+       
+    async def get_accuracy(
+        self, *, extra: Optional[Dict[str, Any]] = None, timeout: Optional[float] = None, **kwargs
+    ) -> Mapping[str, float]:
+        raise NotImplementedError
+    
+    def read(self):
+        # This loop must run for the same amount of time each loop to accurately complete the complementary filter. 
+        while True:
+            now = time.time() 
+            self.velocity = self.read_angular_velocity()
+            self.compass_heading = self.calculate_compass_heading()
+            self.acceleration = self.get_acceleration()
+            self.orientation = self.calculate_orientation()
+            elapsed = time.time() - now
+            time.sleep(dt - elapsed) # full iteration takes 80 ms
+
+    def calculate_compass_heading(self):
+        raw_data = self.i2cbus.read_i2c_block_data(constants.MAG_ADDRESS,  constants.OUT_X_L, 6)
+        values = utils.parse_output(raw_data)
+        if self.calibrate:
+            # hard iron distortion correction
+            values.x -= (self.hard_iron_x_min + self.hard_iron_x_max) /2
+            values.y -= (self.hard_iron_y_min + self.hard_iron_y_max) /2
+
+            # #soft iron distortion correction 
+            scale_y = self.soft_iron_y_max - self.soft_iron_y_min
+            scale_x = self.soft_iron_x_max - self.soft_iron_x_min 
+            ratio = scale_y / scale_x
+            values.x *= ratio
+            values.y *= ratio
+        
+        # tilt compensation.
+        values.x = values.x * math.cos(self.euler_angles['pitch']) + values.z * math.sin(self.euler_angles['pitch'])
+
+        values.y = values.x * math.sin(self.euler_angles['roll'])  * math.sin(self.euler_angles['pitch']) + values.y * math.cos(self.euler_angles['roll']) 
+        - values.z * math.sin(self.euler_angles['roll']) * math.cos(self.euler_angles['pitch'])
+
+        heading = 180 * math.atan2(values.y, values.x) / math.pi
+        if heading < 0:
+              heading += 360
+
+        return heading
+        
+    def get_acceleration(self):
+       raw_data = self.i2cbus.read_i2c_block_data(constants.AG_ADDRESS, constants.OUTX_L_XL, 6)
+       self.accelerometer_raw = utils.parse_output(raw_data)
+      
+       #convert the raw readings to acceleration in m/s^2
+       self.acceleration.x =  self.accelerometer_raw.x * constants.A_GAIN / 1000 * 9.8065
```

Comment:
is the gravity value based on a s specific coordinate in the world?

Reply:
I used the same value that we are using in the witIMU driver, so I think this is for NYC. can add a comment